docs/flip/1009-optimizer-api.md

```diff
@@ -0,0 +1,529 @@
+# flax/opt/base.py
+
+Params = Any
+Gradients = Any
+OptimizerState = NamedTuple
+
+
+class Optimizer(NamedTuple):
+  init: Callable[[Params], OptimizerState]
+  update: Callable[
+      [Gradients, Params, OptimizerState], Tuple[Params, OptimizerState]]
+```
+
+For example, to implement `flax.optim.Momentum` in this new interface:
+
+```python
+# flax/opt/momentum.py
+
+class MomentumState(OptimizerState):
+  momentum: PyTree
+
+
+def momentum(learning_rate: float, beta: float = 0.9) -> Optimizer:
+
+  def init_fn(params: Params) -> OptimizerState:
+    return MomentumState(momentum=jax.tree_map(jnp.zeros_like, params))
+
+  def update_fn(
+    grads: Gradients, params: Params, state: OptimizerState
+  ) -> Tuple[Params, OptimizerState]:
+    new_momentum = jax.tree_multimap(
+        lambda momentum, grad: momentum * beta + grad,
+        opt_state.momentum, grads)
+    new_params = jax.tree_multimap(
+        lambda param, grad: param - learning_rate * grad,
+        params, grads)
+    return new_params, MomentumState(new_momentum)
+
+  return Optimizer(init_fn, update_fn)
+```
+
+Note that unlike in the [previous API], the function `update_fn()` will be
+called directly in the train loop and must take care of processing the pytrees
+of parameters and optimizer state with a `jax.tree_multimap()`. For more
+complex optimizers this can become a hassle but we can provide a small
+functional helper that takes care of this:
+
+```python
+  def update_fn(
+    grads: Gradients, params: Params, state: OptimizerState
+  ) -> Tuple[Params, OptimizerState]:
+
+    @multimap
+    def inner(param, grad, momentum):
+      new_momentum = momentum * beta + grad
+      new_param = param - learning_rate * grad
+      return new_param, new_momentum
+
+    new_params, new_momentum = inner(params, grads, state.momentum)
+    return new_params, MomentumState(new_momentum)
+```
+
+Remarks:
+
+- We'll call `Optimizer` instances from the functional API `optim`, and refer
+  to `flax.optim.Optimizer` instances from the [previous API] as `optimizer`
+  for the remainder of this document.
+- As opposed to the [previous API], this API does not have the option to update
+  hyper parameters. When [using Optax], updating the hyper parameters is instead
+  done by defining a schedule that takes care of updating the hyper parameters
+  during training.
+- Using this simple API directly makes the user code a bit more verbose than
+  using the [previous API], but the difference is smaller with Linen. If you
+  care about a boilerplate-free training loop see the proposed [Linen helper]
+  further down.
+
+## Using Optax
+[Using Optax]: #using-optax
+
+Optax is centered on the idea of composable gradient transformations. Since
+above [functional API] is based on Optax's interface it is straightforward to
+wrap an Optax transformation to comply with the `Optimizer` interface:
+
+```python
+import optax
+
+
+def optax_optimizer(tx: optax.GradientTransformation) -> Optimizer:
```

Comment:
It has the advantage that we adapt our code using this optimizer abstraction and then use any optimizer (including previous `flax.optim.Optimizer`) - but it adds another wrapper...

Adding this to Optax also SGTM, but it litterally saves a single line for the price of an additional indirection - also, Optax is built around the gradient transformation idea, so I can imagine they would prefer to keep this visible in the user code.

Leaving open for discussion...

Reply:
Can you explain what you mean by "It has the advantage that we adapt our code using this optimizer abstraction and then use any optimizer (including previous flax.optim.Optimizer)"? I couldn't follow. Are you saying that it makes the transition from our existing examples easier?

Reply:
It does not make the transition easier, but it would allow us to refactor our existing code once with the new optimizer abstraction and then we can simply exchange one optimizer implementation with another (e.g. `flax.optim.Adam` vs. `optax.adam`) by changing the initialization.

If we decide to update our existing code with the optax abstraction directly (i.e. passing around a `tx` transformation and `opt_state` and then calling `tx.update()` and `optax.apply_updates()`) then we'll have to update the code in multiple places if we want to switch from one implementation to another.

I can see advantages for both approaches and it's really about how much we value one less redirection vs. some more flaxibility.

Reply:
@andsteing did you write "flaxibility" on purpose? 😄 

> It does not make the transition easier, but it would allow us to refactor our existing code once with the new optimizer abstraction and then we can simply exchange one optimizer implementation with another (e.g. flax.optim.Adam vs. optax.adam) by changing the initialization.

That is true, but does it mean we will always have the additional abstraction, even if we completely switch to Optax? I think that in the end, if we like Optax, there is no need to have our own optimizers, and then having an additional abstraction may only make things more complicated.

Reply:
We wouldn't have to use it in that case. But it would allow us to easily replace an Optax optimizer with another optimizer function that transforms `(params, opt_state) -> (params, opt_state)`.

I can see Avital & you don't like it, so I've removed that abstraction completely and go "full Optax" in the FLIP.